marklogic-data-hub/src/test/java/com/marklogic/hub/hubcentral/migration/AllStepsReferencedTest.java

```diff
@@ -0,0 +1,429 @@
+import com.marklogic.hub.impl.FlowManagerImpl;
+import com.marklogic.hub.impl.MappingManagerImpl;
+import com.marklogic.hub.mapping.Mapping;
+import com.marklogic.hub.step.RunStepResponse;
+import com.marklogic.hub.step.StepDefinition;
+import org.apache.commons.io.FileUtils;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.ClassPathResource;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AllStepsReferencedTest extends AbstractHubCoreTest {
+
+    ObjectMapper mapper = new ObjectMapper();
+
+    @Autowired
+    FlowRunner flowRunner;
+
+    @BeforeEach
+    void setUp() {
+        String folderInClasspath = "test-projects/all-steps-referenced";
+        installProjectInFolder(folderInClasspath);
+
+        // copy input directory to ye-olde-project
+        try {
+            File testProjectDir = new ClassPathResource(folderInClasspath).getFile();
+            HubProject hubProject = getHubConfig().getHubProject();
+            File inputDir = new File(testProjectDir, "input");
+            File copyToDir = new File(hubProject.getProjectDir().toFile(), "input");
+            if (inputDir.exists()) {
+                FileUtils.copyDirectory(inputDir, copyToDir);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to load project files: " + e.getMessage(), e);
+        }
+    }
+
+    @Test
+    void validateMigratedFlowsAndSteps() {
+        HubConfig hubConfig = getHubConfig();
+
+        FlowMigrator flowMigrator = new FlowMigrator(hubConfig);
+        flowMigrator.migrateFlows();
+
+        validateFullyReferencedFlowsAndSteps();
+    }
+
+    @Test
+    void migrateAndRunCompleteFlow() {
+        HubConfig hubConfig = getHubConfig();
+
+        FlowMigrator flowMigrator = new FlowMigrator(hubConfig);
+        flowMigrator.migrateFlows();
+        installUserArtifacts();
+
+        RunFlowResponse flowResponse = flowRunner.runFlow(new FlowInputs("CurateCustomerJSON", "1", "2", "3", "4"));
+        flowRunner.awaitCompletion();
+        RunStepResponse runStepResponse = flowResponse.getStepResponses().get("1");
+        assertTrue(runStepResponse.isSuccess(), "Ingestion step failed!");
+        runStepResponse = flowResponse.getStepResponses().get("2");
+        assertTrue(runStepResponse.isSuccess(), "Mapping step failed!");
+        runStepResponse = flowResponse.getStepResponses().get("3");
+        assertTrue(runStepResponse.isSuccess(), "Matching step failed!");
+        runStepResponse = flowResponse.getStepResponses().get("4");
+        assertTrue(runStepResponse.isSuccess(), "Merging step failed!");
+
+        assertTrue(getFinalDocCount("merged-customer") == 6,"Expected 6 merged-customer docs");
```

Comment:
There are unfortunately a lot of equality assertions in the DHF codebase that use assertTrue instead of assertEquals - this should be "assertEquals(6, getFinalDocCount(...), "message").

Reply:
Fixed